File: packages/stronghold_core/lib/models/requests/assign_membership_request.dart
/// Request to assign a membership to a user
class AssignMembershipRequest {
  final int userId;
  final int membershipPackageId;
  final double amountPaid;
  final DateTime startDate;
  final DateTime endDate;
  final DateTime paymentDate;

  const AssignMembershipRequest({
    required this.userId,
    required this.membershipPackageId,
    required this.amountPaid,
    required this.startDate,
    required this.endDate,
    required this.paymentDate,
  });

  Map<String, dynamic> toJson() => {
    'userId': userId,
    'membershipPackageId': membershipPackageId,
    'amountPaid': amountPaid,
    'startDate': startDate.toIso8601String(),
    'endDate': endDate.toIso8601String(),
    'paymentDate': paymentDate.toIso8601String(),
  };
}

File: Stronghold/Stronghold.Application/DTOs/Request/AssignMembershipRequest.cs
using System.ComponentModel.DataAnnotations;

namespace Stronghold.Application.DTOs.Request
{
    public class AssignMembershipRequest : IValidatableObject
    {
        [Range(1, int.MaxValue, ErrorMessage = "Korisnik je obavezan.")]
        public int UserId { get; set; }

        [Range(1, int.MaxValue, ErrorMessage = "Paket članarine je obavezan.")]
        public int MembershipPackageId { get; set; }

        [Range(0.01, 10000, ErrorMessage = "Iznos mora biti veći od 0 i manji ili jednak 10000.")]
        public decimal AmountPaid { get; set; }

        [Required(ErrorMessage = "Datum početka je obavezan.")]
        public DateTime StartDate { get; set; }

        [Required(ErrorMessage = "Datum završetka je obavezan.")]
        public DateTime EndDate { get; set; }

        [Required(ErrorMessage = "Datum uplate je obavezan.")]
        public DateTime PaymentDate { get; set; }

        public IEnumerable<ValidationResult> Validate(ValidationContext validationContext)
        {
            if (EndDate.Date < StartDate.Date)
            {
                yield return new ValidationResult("EndDate ne moze biti prije StartDate-a");
            }

            if (PaymentDate.Date < StartDate.Date || PaymentDate.Date > EndDate.Date)
            {
                yield return new ValidationResult("PaymentDate mora biti unutar perioda clanarine.");
            }
        }
    }
}
